Fail to add index for materialized view in share data mode

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE TABLE t0(c0 INT, c1 INT) DUPLICATE KEY(c0) DISTRIBUTED BY HASH(c0) BUCKETS 1 PROPERTIES('replication_num'='1');
2. CREATE MATERIALIZED VIEW mv1 REFRESH IMMEDIATE MANUAL AS SELECT c0 AS k, count(c1) as cnt FROM t0 GROUP BY c0;
3. CREATE INDEX index_cnt ON mv1(cnt) USING BITMAP COMMENT 'index1';

### Expected behavior (Required)
index can be successfully created

### Real behavior (Required)
Indexes cannot be created for materialized views:
```
2024-05-09 10:47:47,320 WARN (stateChangeExecutor|102) [GlobalStateMgr.replayJournal():2262] got interrupt exception or inconsistent exception when replay journal 301687101, will exit, 
com.starrocks.journal.JournalInconsistentException: failed to load journal type 121
	at com.starrocks.persist.EditLog.loadJournal(EditLog.java:1102) ~[starrocks-fe.jar:?]
	at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:2311) ~[starrocks-fe.jar:?]
	at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:2260) ~[starrocks-fe.jar:?]
	at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1248) ~[starrocks-fe.jar:?]
	at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:339) ~[starrocks-fe.jar:?]
	at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:779) ~[starrocks-fe.jar:?]
	at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:103) ~[starrocks-fe.jar:?]
	at com.starrocks.common.util.Daemon.run(Daemon.java:107) ~[starrocks-fe.jar:?]
Caused by: java.lang.ClassCastException: class com.starrocks.lake.LakeMaterializedView cannot be cast to class com.starrocks.lake.LakeTable (com.starrocks.lake.LakeMaterializedView and com.starrocks.lake.LakeTable are in unnamed module of loader 'app')
	at com.starrocks.alter.LakeTableSchemaChangeJob$LockedDatabase.getTable(LakeTableSchemaChangeJob.java:1032) ~[starrocks-fe.jar:?]
	at com.starrocks.alter.LakeTableSchemaChangeJob.replay(LakeTableSchemaChangeJob.java:756) ~[starrocks-fe.jar:?]
	at com.starrocks.alter.AlterHandler.replayAlterJobV2(AlterHandler.java:200) ~[starrocks-fe.jar:?]
	at com.starrocks.persist.EditLog.loadJournal(EditLog.java:813) ~[starrocks-fe.jar:?]
	... 7 more
```

### StarRocks version (Required)
 - 3.3.0-rc01
 - share data mode

--- a/fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java
+++ b/fe/fe-core/src/main/java/com/starrocks/alter/LakeTableSchemaChangeJob.java
@@ -50,7 +50,6 @@ import com.starrocks.common.util.concurrent.MarkedCountDownLatch;
 import com.starrocks.common.util.concurrent.lock.LockType;
 import com.starrocks.common.util.concurrent.lock.Locker;
 import com.starrocks.journal.JournalTask;
-import com.starrocks.lake.LakeTable;
 import com.starrocks.lake.LakeTablet;
 import com.starrocks.lake.StarMgrMetaSyncer;
 import com.starrocks.lake.Utils;
@@ -195,7 +194,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
     }
 
     // REQUIRE: has acquired the exclusive lock of database
-    void addShadowIndexToCatalog(@NotNull LakeTable table, long visibleTxnId) {
+    void addShadowIndexToCatalog(@NotNull OlapTable table, long visibleTxnId) {
         Preconditions.checkState(visibleTxnId != -1);
         for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
             PhysicalPartition partition = table.getPhysicalPartition(partitionId);
@@ -289,7 +288,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         AgentBatchTask batchTask = new AgentBatchTask();
         MarkedCountDownLatch<Long, Long> countDownLatch;
         try (ReadLockedDatabase db = getReadLockedDatabase(dbId)) {
-            LakeTable table = getTableOrThrow(db, tableId);
+            OlapTable table = getTableOrThrow(db, tableId);
             Preconditions.checkState(table.getState() == OlapTable.OlapTableState.SCHEMA_CHANGE);
             MaterializedIndexMeta indexMeta = table.getIndexMetaByIndexId(table.getBaseIndexId());
             numTablets =
@@ -367,7 +366,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
 
         // Add shadow indexes to table.
         try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = getTableOrThrow(db, tableId);
+            OlapTable table = getTableOrThrow(db, tableId);
             Preconditions.checkState(table.getState() == OlapTable.OlapTableState.SCHEMA_CHANGE);
             watershedTxnId = getNextTransactionId();
             addShadowIndexToCatalog(table, watershedTxnId);
@@ -415,7 +414,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         LOG.info("previous transactions are all finished, begin to send schema change tasks. job: {}", jobId);
 
         try (ReadLockedDatabase db = getReadLockedDatabase(dbId)) {
-            LakeTable table = getTableOrThrow(db, tableId);
+            OlapTable table = getTableOrThrow(db, tableId);
             Preconditions.checkState(table.getState() == OlapTable.OlapTableState.SCHEMA_CHANGE);
             for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
                 PhysicalPartition partition = table.getPhysicalPartition(partitionId);
@@ -486,7 +485,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
 
         try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = getTableOrThrow(db, tableId);
+            OlapTable table = getTableOrThrow(db, tableId);
             commitVersionMap = new HashMap<>();
             for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
                 PhysicalPartition partition = table.getPhysicalPartition(partitionId);
@@ -531,7 +530,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         Set<String> modifiedColumns;
         List<MaterializedIndex> droppedIndexes;
         try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = (db != null) ? db.getTable(tableId) : null;
+            OlapTable table = (db != null) ? db.getTable(tableId) : null;
             if (table == null) {
                 LOG.info("database or table been dropped while doing schema change job {}", jobId);
                 return;
@@ -564,7 +563,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         // updating colocation info. it's possible that after edit log above is written, fe crashes,
         // and colocation info will not be updated , but it should be a rare case
         try (ReadLockedDatabase db = getReadLockedDatabase(dbId)) {
-            LakeTable table = (db != null) ? db.getTable(tableId) : null;
+            OlapTable table = (db != null) ? db.getTable(tableId) : null;
             if (table == null) {
                 LOG.info("database or table been dropped before updating colocation info, schema change job {}", jobId);
             } else {
@@ -586,7 +585,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
     // Note: throws AlterCancelException iff the database or table has been dropped.
     boolean readyToPublishVersion() throws AlterCancelException {
         try (ReadLockedDatabase db = getReadLockedDatabase(dbId)) {
-            LakeTable table = getTableOrThrow(db, tableId);
+            OlapTable table = getTableOrThrow(db, tableId);
             for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
                 PhysicalPartition partition = table.getPhysicalPartition(partitionId);
                 Preconditions.checkState(partition != null, partitionId);
@@ -623,7 +622,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
     }
 
-    private Set<String> collectModifiedColumnsForRelatedMVs(@NotNull LakeTable tbl) {
+    private Set<String> collectModifiedColumnsForRelatedMVs(@NotNull OlapTable tbl) {
         if (tbl.getRelatedMaterializedViews().isEmpty()) {
             return Sets.newHashSet();
         }
@@ -638,7 +637,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         return modifiedColumns;
     }
 
-    private void inactiveRelatedMv(Set<String> modifiedColumns, @NotNull LakeTable tbl) {
+    private void inactiveRelatedMv(Set<String> modifiedColumns, @NotNull OlapTable tbl) {
         if (modifiedColumns.isEmpty()) {
             return;
         }
@@ -670,7 +669,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
 
     // Update each Partition's nextVersion.
     // The caller must have acquired the database's exclusive lock.
-    void updateNextVersion(@NotNull LakeTable table) {
+    void updateNextVersion(@NotNull OlapTable table) {
         for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
             PhysicalPartition partition = table.getPhysicalPartition(partitionId);
             long commitVersion = commitVersionMap.get(partitionId);
@@ -681,11 +680,11 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
     }
 
     @NotNull
-    LakeTable getTableOrThrow(@Nullable LockedDatabase db, long tableId) throws AlterCancelException {
+    OlapTable getTableOrThrow(@Nullable LockedDatabase db, long tableId) throws AlterCancelException {
         if (db == null) {
             throw new AlterCancelException("Database does not exist");
         }
-        LakeTable table = db.getTable(tableId);
+        OlapTable table = db.getTable(tableId);
         if (table == null) {
             throw new AlterCancelException("Table does not exist. tableId=" + tableId);
         }
@@ -728,7 +727,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
 
         try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = (db != null) ? db.getTable(tableId) : null;
+            OlapTable table = (db != null) ? db.getTable(tableId) : null;
             if (table == null) {
                 return; // do nothing if the table has been dropped.
             }
@@ -751,7 +750,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
     }
 
-    void addTabletToTabletInvertedIndex(@NotNull LakeTable table) {
+    void addTabletToTabletInvertedIndex(@NotNull OlapTable table) {
         TabletInvertedIndex invertedIndex = GlobalStateMgr.getCurrentState().getTabletInvertedIndex();
         for (Table.Cell<Long, Long, MaterializedIndex> cell : physicalPartitionIndexMap.cellSet()) {
             Long partitionId = cell.getRowKey();
@@ -769,7 +768,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
     }
 
-    void removeShadowIndex(@NotNull LakeTable table) {
+    void removeShadowIndex(@NotNull OlapTable table) {
         for (long partitionId : physicalPartitionIndexMap.rowKeySet()) {
             PhysicalPartition partition = table.getPhysicalPartition(partitionId);
             Preconditions.checkNotNull(partition, partitionId);
@@ -794,7 +793,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
     }
 
     @NotNull
-    List<MaterializedIndex> visualiseShadowIndex(@NotNull LakeTable table) {
+    List<MaterializedIndex> visualiseShadowIndex(@NotNull OlapTable table) {
         List<MaterializedIndex> droppedIndexes = new ArrayList<>();
         TabletInvertedIndex invertedIndex = GlobalStateMgr.getCurrentState().getTabletInvertedIndex();
 
@@ -903,7 +902,7 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         }
 
         try (WriteLockedDatabase db = getWriteLockedDatabase(dbId)) {
-            LakeTable table = (db != null) ? db.getTable(tableId) : null;
+            OlapTable table = (db != null) ? db.getTable(tableId) : null;
             if (table != null) {
                 removeShadowIndex(table);
             }
@@ -1014,8 +1013,8 @@ public class LakeTableSchemaChangeJob extends AlterJobV2 {
         abstract void unlock(Database db);
 
         @Nullable
-        LakeTable getTable(long tableId) {
-            return (LakeTable) db.getTable(tableId);
+        OlapTable getTable(long tableId) {
+            return (OlapTable) db.getTable(tableId);
         }
 
         @Override

--- a/fe/fe-core/src/main/java/com/starrocks/task/AlterReplicaTask.java
+++ b/fe/fe-core/src/main/java/com/starrocks/task/AlterReplicaTask.java
@@ -317,7 +317,7 @@ public class AlterReplicaTask extends AgentTask implements Runnable {
             }
             Tablet tablet = index.getTablet(getTabletId());
             Preconditions.checkNotNull(tablet, getTabletId());
-            if (!tbl.isCloudNativeTable()) {
+            if (!tbl.isCloudNativeTableOrMaterializedView()) {
                 Replica replica = ((LocalTablet) tablet).getReplicaById(getNewReplicaId());
                 if (replica == null) {
                     throw new MetaNotFoundException("replica " + getNewReplicaId() + " does not exist");

--- a/test/sql/test_alter_mv/R/test_alter_mv
+++ b/test/sql/test_alter_mv/R/test_alter_mv
@@ -1,0 +1,26 @@
+-- name: test_alter_mv_add_index
+CREATE TABLE t0(c0 INT, c1 INT) DUPLICATE KEY(c0) DISTRIBUTED BY HASH(c0) BUCKETS 1 PROPERTIES('replication_num'='1');
+-- result:
+-- !result
+INSERT INTO t0 VALUES(1,1),(2,2),(1,3),(2,4);
+-- result:
+-- !result
+CREATE MATERIALIZED VIEW mv1 REFRESH IMMEDIATE MANUAL AS SELECT c0 AS k, count(c1) as cnt FROM t0 GROUP BY c0;
+-- result:
+-- !result
+CREATE INDEX index_cnt ON mv1(cnt) USING BITMAP COMMENT 'index1';
+-- result:
+-- !result
+function: wait_alter_table_finish()
+-- result:
+None
+-- !result
+[UC] REFRESH MATERIALIZED VIEW mv1 FORCE WITH SYNC MODE;
+-- result:
+73ce5143-1296-11ef-96c0-469314b28ba2
+-- !result
+SELECT k, cnt FROM mv1 ORDER BY k;
+-- result:
+1	2
+2	2
+-- !result

--- a/test/sql/test_alter_mv/T/test_alter_mv
+++ b/test/sql/test_alter_mv/T/test_alter_mv
@@ -1,0 +1,8 @@
+-- name: test_alter_mv_add_index
+CREATE TABLE t0(c0 INT, c1 INT) DUPLICATE KEY(c0) DISTRIBUTED BY HASH(c0) BUCKETS 1 PROPERTIES('replication_num'='1');
+INSERT INTO t0 VALUES(1,1),(2,2),(1,3),(2,4);
+CREATE MATERIALIZED VIEW mv1 REFRESH IMMEDIATE MANUAL AS SELECT c0 AS k, count(c1) as cnt FROM t0 GROUP BY c0;
+CREATE INDEX index_cnt ON mv1(cnt) USING BITMAP COMMENT 'index1';
+function: wait_alter_table_finish()
+[UC] REFRESH MATERIALIZED VIEW mv1 FORCE WITH SYNC MODE;
+SELECT k, cnt FROM mv1 ORDER BY k;

